std/algorithm/searching.d

```diff
@@ -1361,7 +1361,20 @@ if (isInputRange!InputRange &&
 
     alias EType  = ElementType!R;
 
-    static if (isNarrowString!R)
+    // If the haystack is a SortedRange we can use binary search to find the needle.
+    // Works only for the default find predicate and any SortedRange predicate.
+    // 8829 enhancement
+    import std.range: SortedRange;
+    static if (is(InputRange : SortedRange!TT, TT) && isDefaultPred)
+    {
+        auto lb = haystack.lowerBound(needle);
+        if ((lb.length == 0 && haystack[0] != needle) || lb.length == haystack.length
+           || haystack[lb.length] != needle)
+            return haystack[$ .. $];
```

Comment:
This expression is convoluted:

```d
(lb.length == 0 && haystack[0] != needle) || lb.length == haystack.length
           || haystack[lb.length] != needle
```

Simplified:

```d
(x == 0 && haystack[x] != needle) || x == y || haystack[x] != needle
```

Reordered:

```d
x == y || (x == 0 && haystack[x] != needle) || haystack[x] != needle
```

And this exposes the redundancy:

```d
x == y || haystack[x] != needle
```

In fact the more complicated condition had a bug - you're accessing haystack[0] without protection so you'll have a bounds check failure if the haystack is empty.